--- libc/nss/nss_files/files-XXX.c	2011-09-02 22:50:24.000000000 -0400
+++ ../eglibc-2_13.patched/libc/nss/nss_files/files-XXX.c	2012-08-20 21:28:36.545330249 -0400
@@ -40,7 +40,8 @@
 
 #define ENTNAME_r	CONCAT(ENTNAME,_r)
 
-#define DATAFILE	"/etc/" DATABASE
+#define SYSTEM_DATAFILE	"/lib/" DATABASE
+#define CONFIG_DATAFILE	"/etc/" DATABASE
 
 #ifdef NEED_H_ERRNO
 # include <netdb.h>
@@ -64,6 +65,7 @@
 
 /* Maintenance of the shared stream open on the database file.  */
 
+static int parsing_system_data;
 static FILE *stream;
 static fpos_t position;
 static enum { nouse, getent, getby } last_use;
@@ -77,10 +79,18 @@
 
   if (stream == NULL)
     {
-      stream = fopen (DATAFILE, "re");
+      /* Check whether we should open the system (/lib) file or the
+       * one in /etc.
+       */
+      if (parsing_system_data)
+	stream = fopen (SYSTEM_DATAFILE, "re");
+      else
+	stream = fopen (CONFIG_DATAFILE, "re");
 
       if (stream == NULL)
-	status = errno == EAGAIN ? NSS_STATUS_TRYAGAIN : NSS_STATUS_UNAVAIL;
+	{
+	  status = errno == EAGAIN ? NSS_STATUS_TRYAGAIN : NSS_STATUS_UNAVAIL;
+	}
       else
 	{
 #if !defined O_CLOEXEC || !defined __ASSUME_O_CLOEXEC
@@ -175,6 +185,7 @@
 
   /* Reset STAYOPEN flag.  */
   keep_stream = 0;
+  parsing_system_data = 0;
 
   __libc_lock_unlock (lock);
 
@@ -252,6 +263,8 @@
 
   __libc_lock_lock (lock);
 
+again:
+
   /* Be prepared that the set*ent function was not called before.  */
   if (stream == NULL)
     {
@@ -291,11 +304,30 @@
 	     operation (perhaps the buffer was too small).  */
 	  if (status == NSS_STATUS_SUCCESS)
 	    fgetpos (stream, &position);
+	  else if (status == NSS_STATUS_NOTFOUND)
+	    {
+	      if (!parsing_system_data)
+		{
+		  internal_endent ();
+		  parsing_system_data = 1;
+		  goto again;
+		}
+	      else
+		{
+		  parsing_system_data = 0;
+		  last_use = nouse;
+		}
+	    }
 	  else
 	    /* We must make sure we reposition the stream the next call.  */
 	    last_use = nouse;
 	}
     }
+  else if (parsing_system_data)
+    {
+      internal_endent ();
+      parsing_system_data = 0;
+    }
 
   __libc_lock_unlock (lock);
 
@@ -324,6 +356,8 @@
 									      \
   __libc_lock_lock (lock);						      \
 									      \
+again:									      \
+									      \
   /* Reset file pointer to beginning or open file.  */			      \
   status = internal_setent (keep_stream);				      \
 									      \
@@ -341,6 +375,17 @@
 	internal_endent ();						      \
     }									      \
 									      \
+  if (status != NSS_STATUS_SUCCESS && !parsing_system_data)		      \
+    {									      \
+      internal_endent ();						      \
+      parsing_system_data = 1;						      \
+      goto again;							      \
+    }									      \
+  else if (parsing_system_data)						      \
+    {									      \
+      parsing_system_data = 0;						      \
+    }									      \
+									      \
   __libc_lock_unlock (lock);						      \
 									      \
   return status;							      \
