From 9823ced140c3f0142546df8ef1cc24277b723380 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Thu, 26 Jan 2012 16:42:24 -0500
Subject: [PATCH] tinylogin: Move /etc/passwd to /var/passwd

---
 libbb/messages.c |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/libbb/messages.c b/libbb/messages.c
index 42bc5ed..8da6464 100644
--- a/libbb/messages.c
+++ b/libbb/messages.c
@@ -83,22 +83,22 @@ const char * const nologin_file = NOLOGIN_FILE;
 #endif
 
 #ifdef L_passwd_file
-#define PASSWD_FILE        "/etc/passwd"
+#define PASSWD_FILE        "/var/passwd"
 const char * const passwd_file = PASSWD_FILE;
 #endif
 
 #ifdef L_shadow_file
-#define SHADOW_FILE        "/etc/shadow"
+#define SHADOW_FILE        "/var/shadow"
 const char * const shadow_file = SHADOW_FILE;
 #endif
 
 #ifdef L_gshadow_file
-#define GSHADOW_FILE       "/etc/gshadow"
+#define GSHADOW_FILE       "/var/gshadow"
 const char * const gshadow_file = GSHADOW_FILE;
 #endif
 
 #ifdef L_group_file
-#define GROUP_FILE         "/etc/group"
+#define GROUP_FILE         "/var/group"
 const char * const group_file = GROUP_FILE;
 #endif
 
-- 
1.7.6.5

