From 6c33df106d4202062f2fe12dfd9c682991c33561 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Mon, 30 Jan 2012 17:00:40 -0500
Subject: [PATCH] lckpwdf: Also lock files in /var

---
 modules/pam_unix/lckpwdf.-c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/modules/pam_unix/lckpwdf.-c b/modules/pam_unix/lckpwdf.-c
index 7145617..8237854 100644
--- a/modules/pam_unix/lckpwdf.-c
+++ b/modules/pam_unix/lckpwdf.-c
@@ -30,7 +30,7 @@
 #include <selinux/selinux.h>
 #endif
 
-#define LOCKFILE "/etc/.pwd.lock"
+#define LOCKFILE "/var/.pwd.lock"
 #define TIMEOUT 15
 
 static int lockfd = -1;
@@ -76,7 +76,7 @@ static int lckpwdf(void)
 			security_context_t create_context;
 			int rc;
 
-			if(getfilecon("/etc/passwd", &create_context))
+			if(getfilecon("/var/passwd", &create_context))
 				return -1;
 			rc = setfscreatecon(create_context);
 			freecon(create_context);
-- 
1.7.7

