From f6e9c26e5a8cf5c082bb6a58534e4d1c1fb851e0 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Thu, 26 Jan 2012 15:24:00 -0500
Subject: [PATCH] pam_unix: Move /etc/passwd to /var/passwd

---
 modules/pam_unix/passverify.c |   24 +++++++++++++-----------
 1 files changed, 13 insertions(+), 11 deletions(-)

diff --git a/modules/pam_unix/passverify.c b/modules/pam_unix/passverify.c
index 5199a69..ac72bf5 100644
--- a/modules/pam_unix/passverify.c
+++ b/modules/pam_unix/passverify.c
@@ -29,6 +29,8 @@
 #include "bigcrypt.h"
 #include "passverify.h"
 
+#define PASSWD_DIR "/var"
+
 #ifdef WITH_SELINUX
 #include <selinux/selinux.h>
 #define SELINUX_ENABLED is_selinux_enabled()>0
@@ -286,8 +288,8 @@ PAMH_ARG_DECL(int check_shadow_expiry,
 
 /* passwd/salt conversion macros */
 
-#define PW_TMPFILE              "/etc/npasswd"
-#define SH_TMPFILE              "/etc/nshadow"
+#define PW_TMPFILE              PASSWD_DIR "/npasswd"
+#define SH_TMPFILE              PASSWD_DIR "/nshadow"
 #define OPW_TMPFILE             "/etc/security/nopasswd"
 
 /*
@@ -445,7 +447,7 @@ unix_selinux_confined(void)
 {
     static int confined = -1;
     int fd;
-    char tempfile[]="/etc/.pwdXXXXXX";
+    char tempfile[]= PASSWD_DIR "/.pwdXXXXXX";
 
     if (confined != -1)
     	return confined;
@@ -457,7 +459,7 @@ unix_selinux_confined(void)
     }
 
     /* let's try opening shadow read only */
-    if ((fd=open("/etc/shadow", O_RDONLY)) != -1) {
+    if ((fd=open(PASSWD_DIR "/shadow", O_RDONLY)) != -1) {
         close(fd);
         confined = 0;
         return confined;
@@ -579,7 +581,7 @@ save_old_password(pam_handle_t *pamh, const char *forwho, const char *oldpass,
 #ifdef WITH_SELINUX
     if (SELINUX_ENABLED) {
       security_context_t passwd_context=NULL;
-      if (getfilecon("/etc/passwd",&passwd_context)<0) {
+      if (getfilecon(PASSWD_DIR "/passwd",&passwd_context)<0) {
         return PAM_AUTHTOK_ERR;
       };
       if (getfscreatecon(&prev_context)<0) {
@@ -729,7 +731,7 @@ PAMH_ARG_DECL(int unix_update_passwd,
 #ifdef WITH_SELINUX
     if (SELINUX_ENABLED) {
       security_context_t passwd_context=NULL;
-      if (getfilecon("/etc/passwd",&passwd_context)<0) {
+      if (getfilecon(PASSWD_DIR "/passwd",&passwd_context)<0) {
 	return PAM_AUTHTOK_ERR;
       };
       if (getfscreatecon(&prev_context)<0) {
@@ -751,7 +753,7 @@ PAMH_ARG_DECL(int unix_update_passwd,
       goto done;
     }
 
-    opwfile = fopen("/etc/passwd", "r");
+    opwfile = fopen(PASSWD_DIR "/passwd", "r");
     if (opwfile == NULL) {
 	fclose(pwfile);
 	err = 1;
@@ -812,7 +814,7 @@ PAMH_ARG_DECL(int unix_update_passwd,
 
 done:
     if (!err) {
-	if (!rename(PW_TMPFILE, "/etc/passwd"))
+	if (!rename(PW_TMPFILE, PASSWD_DIR "/passwd"))
 	    pam_syslog(pamh,
 		LOG_NOTICE, "password changed for %s", forwho);
 	else
@@ -854,7 +856,7 @@ PAMH_ARG_DECL(int unix_update_shadow,
 #ifdef WITH_SELINUX
     if (SELINUX_ENABLED) {
       security_context_t shadow_context=NULL;
-      if (getfilecon("/etc/shadow",&shadow_context)<0) {
+      if (getfilecon(PASSWD_DIR "/shadow",&shadow_context)<0) {
 	return PAM_AUTHTOK_ERR;
       };
       if (getfscreatecon(&prev_context)<0) {
@@ -876,7 +878,7 @@ PAMH_ARG_DECL(int unix_update_shadow,
 	goto done;
     }
 
-    opwfile = fopen("/etc/shadow", "r");
+    opwfile = fopen(PASSWD_DIR "/shadow", "r");
     if (opwfile == NULL) {
 	fclose(pwfile);
 	err = 1;
@@ -955,7 +957,7 @@ PAMH_ARG_DECL(int unix_update_shadow,
 
  done:
     if (!err) {
-	if (!rename(SH_TMPFILE, "/etc/shadow"))
+	if (!rename(SH_TMPFILE, PASSWD_DIR "/shadow"))
 	    pam_syslog(pamh,
 		LOG_NOTICE, "password changed for %s", forwho);
 	else
-- 
1.7.6.5

